src/lib.rs

```diff
@@ -4,4 +4,161 @@
 //! Supports the [ESP32-WROOM-32E](https://www.espressif.com/sites/default/files/documentation/esp32-wroom-32e_esp32-wroom-32ue_datasheet_en.pdf), [ESP32-WROOM-32UE](https://www.espressif.com/sites/default/files/documentation/esp32-wroom-32e_esp32-wroom-32ue_datasheet_en.pdf) modules.
 //! Future implementations will support the [ESP32-WROOM-DA](https://www.espressif.com/sites/default/files/documentation/esp32-wroom-da_datasheet_en.pdf) module.
 //! 
-//! NOTE This crate is still under active development. This API will remain volatile until 1.0.0
\ No newline at end of file
+//! NOTE This crate is still under active development. This API will remain volatile until 1.0.0
+
+
+// This is just a placeholder for now. We will need to implement a socket struct or
+// use an existing crate
+type Socket = [u8; 1];
+
+// TODO: Find a way to share these between NinaCommandHandlers
+const START_CLIENT_TCP: u8 = 0x2du8;
+const GET_FW_VERSION: u8 = 0x37u8;
+
+pub struct Wifi<C: NinaCommandHandler<I>> {
+  command_handler: C,
+}
+
+pub struct FirmwareVersion {
+    major: u8,
+    minor: u8,
+    patch: u8
+}
+
+impl FirmwareVersion {
+    fn new(version: [u8; 5]) -> FirmwareVersion {
+        FirmwareVersion::parse(version)
+    }
+
+    // Takes in 5 bytes (e.g. 1.7.4) and returns a FirmwareVersion instance
+    fn parse(version: [u8; 5]) -> FirmwareVersion {
+        // TODO: real implementation
+        FirmwareVersion {
+            major: 1,
+            minor: 7,
+            patch: 4
+        }
+    }
+}
+
+impl<I: IoInterface, C: NinaCommandHandler<I>> Wifi<I, C> {
+    // fn connect(&self) -> Result<T> {
+    //     self.command_handler.start_client_tcp()
+    // }
+
+    fn get_firmware_version(&self) -> Result<FirmwareVersion, Error> {
+      self.command_handler.get_fw_version()
+    }
+}
+
+impl SpiCommandHandler<I> {
+    fn send_command(command: u8, parameters: [u8; 5]) -> Result<FirmwareVersion, Error> {
+        Ok(FirmwareVersion::new([1,0,7,0,4]))
+      }
+}
+
+impl<I> NinaCommandHandler<I> for SpiCommandHandler<I> {
+
+    type Error = dyn std::error::Error;
+
+    fn start_client_tcp(&self, socket: Socket) -> Result<FirmwareVersion, Error> {
+        // TODO: implement a trait interface and set of structs for different parameter sets, e.g. SocketType
+        SpiCommandHandler::send_command(START_CLIENT_TCP, [socket])
```

Comment:
I know we discussed not over-complicating this for a first pass, but I think we could create a very simple trait that send_command then takes as a struct parameter type for each parameter set type. The trait would then define a `into_bytes()` function or something like it, much like the [Heapless String](https://docs.rs/heapless/latest/heapless/struct.String.html#method.into_bytes) does. I think this will make our interface stronger and actually simpler to do it this way. The goal being to always get parameters down into it's array of bytes form for `send_command()` to send out over the particular data bus instance.

Reply:
The placeholder totally works and this comment thread will be here as a reminder (along with the inline code comment) to take a second pass on it later.